posts/inside-rust/2019-12-17-bisecting-rust-compiler.md

```diff
@@ -0,0 +1,158 @@
+---
+layout: post
+title: "Bisecting Rust Compiler Regressions with cargo-bisect-rustc"
+author: Santiago Pastorino
+team: the compiler team <https://www.rust-lang.org/governance/teams/compiler>
+---
+
+Let's say that you've just updated the Rust compiler version and have
+tried to compile your application and see a failure that wasn't there
+before. That's likely due to a regression in the compiler.  We've just
+released
+[`cargo-bisect-rustc`](https://github.com/rust-lang/cargo-bisect-rustc),
+this tool makes it super easy to find exactly when behavior has
+regressed in rustc.
+
+`cargo-bisect-rustc` automatically downloads rustc artifacts and tests
+them against a project you provide until it finds the regression. At
+minimum, it will identify the nightly release which triggered the
+regression; but if the regression occurred in the last 90 days, it will
+even figure out the exact PR, which is often very useful in helping us
+fix the problem.
+
+`cargo-bisect-rustc` was created originally by Mark Rousskov. I extended
+it recently to make it easier to use.
+
+To install the tool run:
+
+```sh
+cargo install cargo-bisect-rustc
+```
+
+## Finding a regression
+
+We are going to use [this "old" reported rustc
+regression](https://github.com/rust-lang/rust/issues/64945) as an
+example:
+
+Our application consist only of this file:
+
+```rust
+pub struct Slice<'a, T>(&'a [T]);
+
+impl<'a, T: 'a> Slice<'a, T> {
+    pub const EMPTY: Self = Slice ({
+        let v: &[T] = &[];
+        v
+    });
+}
+
+fn main() {
+    let s = Slice(&[1, 2]);
+    assert!(s.0 != Slice::EMPTY.0);
+}
+```
+
+Then we run `cargo bisect-rustc --end=2019-10-01`.
```

Comment:
Why `2019-10-01`? Maybe say something like "Since this bug was fixed on yyyy-mm-dd, we're using `2019-10-01`' in the next paragraph.

Reply:
Fixed.